File: problem_4.cpp
#include <chrono>
#include <iostream>
#include <functional>

bool hasThreeDigitDivisor (long n) 
{
    for (int i = 999; i >= 100; i--) 
    {
        if (n % i == 0 && n / i >= 100 && n / i < 1000)
        {
            std::cout << "Divisors: " << i << "&" << (n / i) << std::endl;
            return true;
        }
    }
    return false;
}

const int max_digit = 9;
int digits[3] = { max_digit, max_digit, max_digit };

std::function<long()> generateNextPalindrome = [&digits] () 
{
    long toRet = long(digits[0]) * long(1e5 + 1) + long(digits[1]) * long(1e4 + 10) + long(digits[2]) * long(1e3 + 1e2);
    if (digits[2] > 0)
    {
        (digits[2])--;
    }
    else if (digits[1] > 0) 
    {
        (digits[2]) = max_digit;
        (digits[1])--;
    }
    else if (digits[0] > 0) 
    {
        (digits[1]) = max_digit;
        (digits[0])--;
    }
    else
    {
        return (long)0;
    }
    return toRet;
};

int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();

    long palindromeToTest = 1e6;
    do {
        palindromeToTest = generateNextPalindrome();
        if (hasThreeDigitDivisor(palindromeToTest))
        {
            break;
        }
    }
    while (palindromeToTest != 0);

    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << palindromeToTest << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_4.rs
use std::time::{Instant};

const MAX_DIGIT:usize = 9;

fn has_three_digit_divisor (n: usize) -> bool
{
    for i in (100..1000).rev()
    {
        if n % i == 0 && n / i >= 100 && n / i < 1000 
        {
            println!("Divisors: {} | {}", i, n/i);
            return true
        }
    }
    false
}


fn main () 
{
    let mut digits: (usize, usize, usize) = (MAX_DIGIT, MAX_DIGIT, MAX_DIGIT);
    let mut generate_next_palindrome = || -> usize 
    {
        let to_ret : usize = digits.0 * (100000 + 1) + digits.1 * (10000 + 10) + digits.2 * (1000 + 100);
        if digits.2 > 0
        {
            digits.2 = digits.2 - 1;
        }
        else if digits.1 > 0
        {
            digits.2 = MAX_DIGIT;
            digits.1 = digits.1 - 1;
        }
        else if digits.0 > 0
        {
            digits.1 = MAX_DIGIT;
            digits.0 = digits.0 - 1;
        }
        else 
        {
            return 0
        }
        return to_ret
    };

    let start = Instant::now();

    let mut palindrome = generate_next_palindrome();
    while palindrome > 0 && !has_three_digit_divisor(palindrome)
    {
        palindrome = generate_next_palindrome();
    }
    let duration = start.elapsed();
    println!("Result: {}", palindrome);
    println!("Timing: {}", duration.as_nanos());
}

